### ISSUE TITLE
Late Fee Calculation Bug

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1189]
In GitLab by @kdjstudios on Oct 10, 2018, 11:26

Hello Team,

This ticket is to fix the calculation for late fees. Currently SAB is not correctly calculating the late fee. I believe the below is the outline of how it should be:

```java
//Do we even need to charge a late fee?
if (late_fee_enabled == true){ 

//Variables...
payment_date = effective_date_of_payment; //Make sure we are using the date a payment is marked as effective, not necessarily the date it was made.
due_ date = current_date; // Will be updated to correct date during process (prevent any Null errors)
charge_late_fee = true; // Error on the side of caution (better to charge and need to credit then not charge)
current_balance = previous_remaining_balance; // This is the "previous balance" amount that would display on the current/new invoice.

//Find the due date
if (net_terms != 0){
     due_date = previous_invoice_date + net_terms; //Add the net terms to the last invoice date
}else{
     due_date = current_invoice_date; // If net terms = 0 then use current invoice date (**Might need to change this**)
}

// Find if we need a late fee
if (payment_date <= due_date) { // Was a payment made on time?
     if (current_balance <= 0) { // Check to see if there is a balance 
          charge_late_fee = false; // Payment was made on time and there is no balance. No late fee
     }else {
          charge_late_fee = true; // Payment was made, but there is still a balance. Yes Late fee
     }
}else{ //Was no payment made?
     if (current_balance <= 0) { // Check to see if there is a balance 
          charge_late_fee = false; // There is/was no balance, so no payment was needed. No late fee
     }else {
          charge_late_fee = true; // No Payment was made, but there is still a balance. Yes Late fee
     }
}

//Log entry? "Account: XX | Late fee enabled: True | Late fee amount: XX.XX"


//Late fee not enabled
}else{ 
    //Log entry? "Account: XX | Late fee enabled: False"
}
//Generate draft invoice
```

**NOTES:** Manual invoices we may want to generate a pop up that says "Please review the draft invoice for any late fees that may not need to be charged."

**Questions:** (The answers to these will effect the calculation above) 

*  Does the net terms include the day of the invoice, or is it in addition to the invoice?

> Example: if I have a invoice date of 1/1/18 and net terms of 15; is the due date the 15th or the 16th?

*  Does the due date include the day of, or is it the day after?

> Example: If I have a due date of 1/15/18; is a payment made on the 15th late?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2018, 11:45

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 10, 2018, 13:31

I just realized that the above would not take into account where there are multiple invoices within a net terms. Sorry. Please ignore the above code. Just a question, when we generate an invoice do we set a due date specific to that invoice at that time? Or are we calculating past due dates when generating new invoices?

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 29, 2018, 06:16

@smasih @kdjstudios I think the scenario Kyle is describing is that we have a draft invoice and then another invoice gets generated in the normal cycle.  Then the following billing cycle, how does it handle the late charges?  

In my discussion w/Sumeet he indicates that it will always take the most recent invoice when calculating dates and fees.  So that invoice would have the accurate outstanding balance and also the same invoice date as any other invoices in that cycle, so the date calcs would be the same.

Kyle, does that address your question?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2018, 07:53

@tim.traylor 

Thank you, but if I am understanding the above explanation then for accounts that have net terms that are larger then the billing cycle period they would never calculate a late fee, since we are only looking at the most recent invoice, right?

Example:

An Account is on a 14 day billing cycle (two weeks) but has a net terms of 30 days. So if we are only looking at the last invoice, then we are not looking at the correct invoice to determine a late fee. Since there would be at least two invoices generated after the invoice that actually has the due date that is needing to be checked. Correct?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 31, 2018, 05:46

@kdjstudios @tim.traylor 

- Tim clarified that there is no tier system for late charges.
- System only charges late fee only on those invoices which are outstanding.

System only charges late fee 

#### Before Change

````ruby
  def late_fee_balance
    if invoices.count == 0
      return 0.to_money
    end
    total_ar_balance = 0.to_money
    if net_terms <= 0
      total_ar_balance = invoices.outstanding.sum(&:balance)
    else
      #invoices = invoices
      #net_terms = invoice.account.net_terms
      net_terms_days_ago = self.current_billing_cycle.invoice_date - self.net_terms.days # Date.today - net_terms.days
      last_invoice = invoices.first
      unless last_invoice.nil?
        last_invoice_date = invoices.first.created_at.to_date
        late_invoices = invoices.where(:finalized_total.gt => 0.to_money).any_of({:paid_date.gt => last_invoice_date, :days_outstanding.gt => net_terms},{:paid_date => nil, :invoice_date.lt => net_terms_days_ago})
        total_ar_balance = late_invoices.sum(&:finalized_amount)
      end
    end
end
````

#### After Change

````ruby
  def late_fee_balance
    if invoices.count == 0
      return 0.to_money
    end
    total_ar_balance = 0.to_money
    if net_terms <= 0
      total_ar_balance = invoices.outstanding.sum(&:balance)
    else
      ## ------------ After change ----------------
      net_terms_days_ago = self.previous_billing_cycle.invoice_date + self.net_terms.days 
      last_invoice = invoices.first
      unless last_invoice.nil?
        last_invoice_date = invoices.first.created_at.to_date
        late_invoices = invoices.where(:finalized_total.gt => 0.to_money).any_of({:paid_date.gt => net_terms_days_ago},{ :paid_date => nil, :invoice_date.lt => net_terms_days_ago})
        total_ar_balance = late_invoices.sum(&:finalized_amount)
      end
    end
end
````

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Oct 31, 2018, 06:11

**Current Process Fails With These Parameters:**
* net terms 60 days
* 7/1 invoice due 9/1 unpaid
* 8/1 invoice due 10/1 unpaid
* 9/1 cycle, prev inv date would be 8/1 w/calculated due date of 10/1



**The new process needs to be as follows:**
```ruby
def late_fee_balance
    if invoices.count == 0
      return 0.to_money
    end
    total_ar_balance = 0.to_money
    if net_terms <= 0
      total_ar_balance = invoices.outstanding.sum(&:balance)
    else
      ## ------------ After change ----------------
for each invoice with an unpaid balance
    late_charge_after = that invoice date + net terms
    if current billing cycle invoice date >= late_charge_after
        amount_to_asses_late_charge_on = that invoice unpaid balance

return amount_to_asses_late_charge_on 

    end
end
```

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2018, 14:37

May we please get an update on the status of this?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Nov 21, 2018, 24:40

@kdjstudios 

I have a question for @tim.traylor on this.

1) I would suggest going with the concept of one invoice date.    
2) How are we going to handle invoices those were after due date?

````ruby
  def late_fee_balance
    #----------------------------------------
    # changes approved in #1191 Gitlab Ticket
    #----------------------------------------
    return 0.to_money if invoices.count <= 0 # return if there are no invoices
    total_ar_balance = 0.0
    total_ar_balance2 = 0.0 #used to aid bad late charge 

    # if net terms are 0 then the due date is invoice to invoice/ invoice should be paid before next invoice date.
    if net_terms <= 0
      total_ar_balance = invoices.outstanding.sum(&:balance)
    else
      # when net term on account are greater than 0       
      #due_date can be different for all invoices.
      late_invoices_with_bal_gt_0 = self.invoices.where(finalized: true, :balance.gt => 0.to_money).asc(:generated_at)
      late_invoices_with_bal_gt_0.each do |inv|
        due_date = inv.invoice_date + net_terms.days
        if self.current_billing_cycle.invoice_date > due_date
          total_ar_balance += inv.finalized_amount.to_f
          total_ar_balance2 += inv.balance.to_f #used to aid bad late charge 
        end
      end
      total_ar_balance = total_ar_balance.to_money
      total_ar_balance2 = total_ar_balance.to_money
    end
    #----------------

````

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Nov 21, 2018, 06:03

Noting our discussion today:

Sumeet realized that the way the late charging works it is only going to late charge for invoices that have a non-zero balance when the new invoices are generated.  It will not account for invoices that were paid, but AFTER the due date.  So, we need to do the following:

* Add a new check that looks for invoices that have a zero balance, but where the payment date that satisfied the invoice was AFTER the due date of the invoice.
* When this code first goes live it would try and go back through all invoices for every account, and it could generate an enormous volume of late charges.  To prevent this we want to add a date check limiter to how far back this check goes.  We will hard code the date that the change is being released and then lookup the billing cycle prior to that date.  We will then not look at any invoices prior to that previous billing cycle date.  
*  Regarding calculating the amount on which late charges are applied, we need to add the payments made AFTER the due date of the invoice to determine what amount we will base late charges on.
*  To confirm Sumeet's question, if the satisfying payment is made ON the due date the account does not get late charged.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 26, 2018, 02:28

@kdjstudios 

We have successfully tested Late fee on the local system and all are of them working.  

Test cases are shown in spreadsheets below.
 https://docs.google.com/spreadsheets/d/1EAy88-ikNnDvz30XxUOQQW3wiZtjqpXIPSqB_wn809g/edit#gid=0

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 13, 2018, 04:30

@kdjstudios 

We have updated the code on the dev server for the testing. We have send the test cases and result in the previous comment (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1191#note_30620)with the spreadsheet. So Please test it.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 13, 2018, 12:44

@amishra @pchaudhary @smasih @tim.traylor 

Great test cases. I was able to run two billing cycles on the Demo site on the Dev server. The first was with no payments and to insure Late charges were assessed. The late charges were assessed correctly. The second, I made payments on all accounts that had a late charges and to insure that late charges would not be assessed after payments. Everything worked correctly. I will attempt a few more specific tests, but so far the late charges appear to be calculating and assessing correctly.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 13, 2018, 13:02

I did a third billing cycle, I made 'late' payments on certain accounts; and some accounts did not make any payments on. I then ran another Billing cycle to see how late charges were assessed. 
 - For account "Example007" its invoice "DEMO-823" should have had a late charge assessed, but it did not. the payment was made over a month late. 
 - For account "DM-AC-7" its invoice "DEMO-841" should have had a late charge assessed, but it did not. the payment was made about 10 days late. 
 - For account "DM-AC-13" its invoice "DEMO-837" should have had a late charge assessed, but it did not. the payment was made about 20 days late. 

Overall it appears 'late' payments are not being calculated correctly. but non-payments are working. 

I will attempt a few more specific tests and post their updates here.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 14, 2018, 06:57

@kdjstudios 

We dug into the code and found some existing bug in the code.
* When the payment was done about 10 days late and invoice amount is 0 then when we creating next invoice we were checking the invoice paid to date (the paid date was the current date).
* When we create two invoices and that invoice pay after the due date and when we creating the third invoice and pay a late charge of the previous invoice and third invoice pay within due date. Then we create the fourth invoice then we paying again late charges of previous invoices which were paid after the due date(Which had already paid in the third invoices.).


So We have to fix those issue and internal testing was left so once we complete the testing I will let you know.

thanks.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Dec 17, 2018, 05:12

@kdjstudios 

We have set up for your testing on the dev server. Please test it and let me know if there is anything to do in our end.

**Test cases & Results**   

**Case 1:**    

 `Note` **Net Terms is 10 and the billing cycle is master.**
 * We have created one invoice "DEMO-1043"  on account "test000004" under "Demo Site" on one billing cycle. After that, we have payment done about two months before and an invoice amount is 0.
 ![image](/uploads/cb79312b4a5bea2d81155fd41a4f2281/image.png)

 * We created another invoice "DEMO-1070" on same account number and site with different billing cycle. 
 There are no late charges. Payment was done about 1.5 months late.
 ![image](/uploads/b998ed08047aa0b5340a14af9b4067e5/image.png)

 * We have created another invoice "DEMO-1107"  and include the previous late charge.
 Payments paid under due date.
  ![image](/uploads/9890fe4fe9df470f147bb74b29d9e75f/image.png)

 * When created another invoice "DEMO-1109" there is no late charge.
  ![image](/uploads/32b26c65b35efb72ce8e3deb0e6c0a3a/image.png)


**Case 2:**
 `Note` **Net term is 45**

 * We have tested on account "test aavaz" inside demo site.(https://dev.sabilling.com/accounts/5be521b8cf85c161d000033e.html?sid=59423ab5cf85c111be000026).
  ![image](/uploads/fb78bb0c3f022edc5ab3dc3d499ad1e3/image.png)



I have one question, As u can see in below screenshot
  ![image](/uploads/e033f848e73a8c5dee7da19754d6da3f/image.png)


There are two invoices "DEMO-1236" and "DEMO-1288". 
 * In first invoice total amount is "**1320.0**"  and we have done payment of "1120". So that the invoice balance will become"200".
 * In second invoice we have late fees of the first invoice is "132.0".

As per current code, the late charges are applying on the total of the invoice, Which is "1320" here.
But according to us, It should be on the remaining balance "1320-1120" which is "200".

We just want to let you know that Is it correct or not?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 17, 2018, 08:41

@amishra @pchaudhary @smasih @tim.traylor 

Thank you for the update. Below are my responses for each test case:

### Test Case One:

Upon looking at Demo-54 account on Demo site on Dev This was the current standing below. It had a total outstanding balance of "$4582.93" due on "12/16/2021".

[invoice_DEMO-1282.pdf](/uploads/630803c0c25d251ba0a35493a99ca370/invoice_DEMO-1282.pdf)

I made a payment for "$4582.93" on "12/26/2021". Which is 10 days late, so we should have an late charge on the next invoice for the full balance due. I ran the next billing cycle and here is the invoice.

[invoice_DEMO-1311.pdf](/uploads/719c77119d99f4a3018cf5af7602ffad/invoice_DEMO-1311.pdf)

**Success**. There is a late payment. 

**FAIL**. The amount of the late fee appears incorrect. 
> I am concerned with the amount. The 'Late Charge Percentage" setup at the site is supposed to be 10%. Which if I am looking at the invoice correctly, then the late charges is not calculating right. We only charged "$271.80"; which is only about 60% of the amount I would have expected to be charged. Shown below. 
> 
> 4582.93 * 0.10 = 458.293
>
> Can someone please help me understand what I am missing or if the calculation is wrong. 



 

### Test Case Two:

> As it would appear is stated in both @smasih comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1191#note_29209 and in @tim.traylor comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1191#note_30343 this should only be applied on the (remaining) outstanding balance. Which in your scenario above would be the 200 amount, that was not paid.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 2, 2019, 04:42

@kdjstudios @tim.traylor 

We suggest adding the **BETA** label again on the Detail Open AR Report until this will be fixed & released.    
Because Detail Open AR is also calculating the Late Fee which would affect the calculation of report.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 7, 2019, 12:46

@smasih @amishra @pchaudhary May we please have an update on status of this ticket. 

Also need to add improvement to calculation to take into account pending credits (staged fees) before we calculate the late fee. #1302

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 8, 2019, 07:19

@kdjstudios 

While development on this ticket, we found some issues.  
1.  If there are more than one late charges on the account & we deactivated any of them then it's calculating wrongly for the late charges.  
1.  The other we found that, if there is any manual payment on the account & not printed on the invoice yet. In the other tab, we create a manual draft invoice in which the payment got printed. There is still an option to delete the manual payment on the other tab which should not be there (*untill we reload the page*).

* We have fixed all of the above issues while development on this issue.

* Most of the things are completed for this particular issue. Now, the only thing remaining is:
  * Currently, the invoice is showing 2 separate line items for the late charges, if there are more than 1 late charges to be applied on the invoice. However, the calculation is accurate. Now, we need to remove multiple line items & put only a single line item for the late charges wih the total charge.

  * Testing needs to be done for multiple scenarios.

* We'll continue tomorrow on fixing the invoice view & testing & will update you.  

####  NOTE

* We need to discuss one of the scenario in which the late charge calculation is calculating incorrectly.

  * Let's say there is a Billing Cycle **A** having an invoice **I-1** of **$ 100**. 
  * Then we create a new Billing Cycle **B** & create an invoice **I-2** of **$ 0**. 
  * Now, if we make a payment of **$ 100** for invoice **I-1**. (After the due date) 
  * If we create a Billing Cycle **C** and create an invoice **I-3** in which there should be the late charge for invoice **I-1** but it'll not be there, becuase we have a limit for going back to only single billing cycle to check late paid invoices.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 8, 2019, 08:31

@pchaudhary @amishra @smasih @tim.traylor 

Currently, the invoice is showing 2 separate line items for the late charges, if there are more than 1 late charges to be applied on the invoice. However, the calculation is accurate. Now, we need to remove multiple line items & put only a single line item for the late charges wih the total charge.

> Who approved this change? I would believe if we have two separate charges, we want to keep them separate. Is that how multiple late charges were displayed in the past? @tim.traylor what is your input?


If we create a Billing Cycle **C** and create an invoice **I-3** in which there should be the late charge for invoice **I-1** but it'll not be there, becuase we have a limit for going back to only single billing cycle to check late paid invoices.

> When did we ever put a limit for going back on billing cycles? I do not recall this discussion, nor can I find it in the comments anywhere.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 8, 2019, 09:44

removed milestone

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 8, 2019, 10:11

This was discussed today and the following was mentioned:

- The current updates are needing to be released on the 1-19 release no matter what to prevent further invalid late charges.
- We are aware of a few concerns/issues that will still need to be addressed over time:
 - Invoices that were paid late but had an zero balance invoice after. (Needs further attention)
 - Detail Open AR (Needs updated with correct calculation #1312 )


@pchaudhary @amishra @smasih 

Please let me know once this is ready to be tested again.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 10, 2019, 05:06

@kdjstudios 

We have completed the development & Testing from our end. 
* You can find the updated test cases sheet on the link - [Test Case Spreadsheet] (https://docs.google.com/spreadsheets/d/1EAy88-ikNnDvz30XxUOQQW3wiZtjqpXIPSqB_wn809g/edit#gid=0)

* We found a new scenario while testing which is failing at the moment.  
  * Let's say there is an invoice **I-1 of $ 100**. 
  * We make a payment of $ 50 with the paid date as the after due date which means there should be a late charge on **$ 100** while creation of a new invoice. 
  * But while calculation of a late charge on the new invoice, it is taking **$ 50 only** of the previous invoice, which should be **$ 100** as the paid date of $ 50 is also after the due date.  

* While development on this issue, we found some other issue regarding the manual payment which got fixed while fixing this issue. We are going to create new tickets for that particular issue.
  * #1320 
  * #1321

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 10, 2019, 13:50

@pchaudhary 

Thank you for the update. Please be sure to review this with @tim.traylor in the morning meeting with him. 

Also this currently has two labels for which server this is on, So I have not a clue which server to run tests on. Please be sure to update to only have one label. Also be sure to keep a 'PL' label on the ticket so we know which status. NOTE: the server labels, are not 'PL' labels.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 11, 2019, 07:44

@kdjstudios 

Sorry for that inconvenience. This is live on Merge2.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 11, 2019, 09:29

@pchaudhary 

I was able to run a few test billing cycles and monitor late fees and everything except for the described above scenarios appear to be generating late fees correctly. 

> As a separate question, how much time would it take to remove all of the billing cycles for the demo site on all dev servers? Right now we are in year 2021 and so entering payments takes forever due to the pop up because I have to post the payments in the future.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 18, 2019, 03:02

@kdjstudios 

We are not going to release this as the label is not allowing us to do.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 18, 2019, 11:08

@pchaudhary @tim.traylor @amishra @smasih @rchauhan @abakshi 

Please if you feel a label is incorrect reach out via the WhatsApp. This was discussed how many times this week? 10+ times? And ever single time it was stated that this needed to be released this week.

This **MUST** be released this week. It has been over 3 months since we found that the late fee calculation were not working. we need this corrected.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 19, 2019, 06:40

@kdjstudios 

We have released this on both of the production servers.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 19, 2019, 22:07

Thank you. All appears to be working as expected.

We will still need to address all of the known bug scenarios with this. I believe they have all been documented in separate tickets correct? If not, please do so.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 21, 2019, 05:42

removed milestone

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 22, 2019, 03:36

@kdjstudios 


Thanks for the update. 

* We also believe they all have been documented in separate tickets.
* There is a failed test case mentioned in the document which is accepted. 

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 22, 2019, 12:29

@pchaudhary 

There is a failed test case mentioned in the document which is accepted.

- Which test case was this? I think it was accepted to allow us to release this but will still need to be addressed in a separate ticket.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 22, 2019, 23:30

@kdjstudios 

Failed Cases - 

##### Accepted

* Let's say there is a Billing Cycle A having an invoice I-1 of $ 100.
* Then we create a new Billing Cycle B & create an invoice I-2 of $ 0.
* Now, if we make a payment of $ 100 for invoice I-1. (After the due date)
* If we create a Billing Cycle C and create an invoice I-3 in which there should be the late charge for invoice I-1 but it'll not be there, because we have a limit for going back to an only single billing cycle to check late paid invoices.


##### New Fail Scenario - 
* Let's say there is an invoice I-1 of $ 100. 
* We make a payment of $ 50 with the paid date as the after due date which means there should be a late charge on $ 100 while creation of a new invoice. 
* But while the calculation of a late charge on the new invoice, it is taking $ 50 only of the previous invoice.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 23, 2019, 08:55

@pchaudhary 

Thank you, I believe both of those scenarios will need to be addressed and resolved. If I recall from the previous conversation those were acceptable due to us needing to fix the MAJOR underlining issue promptly. I was unable to find separate ticket for those, so I will create tickets for both of those and review them with Tim. 

#1349 and #1350

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 23, 2019, 23:27

@kdjstudios 

Thanks for the update. We'll wait for your response on the above mentioned tickets.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 24, 2019, 08:49

closed
